File: app/src/main/java/net/bloople/manga/ReadingSession.java
package net.bloople.manga;

import android.annotation.SuppressLint;
import android.content.Context;

import androidx.fragment.app.FragmentActivity;
import androidx.viewpager2.widget.ViewPager2;

import net.bloople.manga.audit.BooksAuditor;

class ReadingSession {
    private static final int CACHE_PAGES_LIMIT = 5;
    private Context context;
    private ViewPager2 pager;
    private Library library;
    private Book book;
    private BookMetadata metadata;
    private BooksAuditor auditor;

    ReadingSession(Context context, Library library, Book book) {
        this.context = context;
        this.library = library;
        this.book = book;
        metadata = BookMetadata.findOrCreateByBookId(context, book.getId());
        auditor = new BooksAuditor(context);
    }

    void start() {
        metadata.lastOpenedAt(System.currentTimeMillis());
        metadata.openedCount(metadata.openedCount() + 1);
        metadata.save(context);

        auditor.opened(library, book, page());
    }

    @SuppressLint("WrongConstant")
    void bind(FragmentActivity fa, ViewPager2 pager) {
        this.pager = pager;

        pager.setAdapter(new BookPagerAdapter(fa, book));

        pager.registerOnPageChangeCallback(new ViewPager2.OnPageChangeCallback() {
            @Override
            public void onPageSelected(int position) {
                bookmark(position);
            }
        });

        pager.setOffscreenPageLimit(CACHE_PAGES_LIMIT);
    }

    int page() {
        return pager.getCurrentItem();
    }

    void page(int page) {
        pager.setCurrentItem(page, false);
    }

    void go(int change) {
        pager.setCurrentItem(pager.getCurrentItem() + change, false);
    }

    private void bookmark(int page) {
        metadata.lastReadPosition(page);
        metadata.save(context);
    }

    void resume() {
        page(metadata.lastReadPosition());
    }

    void finish() {
        if(page() == book.getPages() - 1) bookmark(0);
        auditor.closed(library, book, page());
    }
}

File: app/src/main/java/net/bloople/manga/ReadingSession.kt
package net.bloople.manga

import androidx.viewpager2.widget.ViewPager2
import net.bloople.manga.audit.BooksAuditor
import android.annotation.SuppressLint
import android.content.Context
import androidx.fragment.app.FragmentActivity
import androidx.viewpager2.widget.ViewPager2.OnPageChangeCallback

class ReadingSession(private val library: Library, private val book: Book) {
    private var pager: ViewPager2? = null
    private val metadata: BookMetadata = BookMetadata.findOrCreateByBookId(book.id)
    private val auditor = BooksAuditor()

    fun start() {
        metadata.lastOpenedAt = System.currentTimeMillis()
        metadata.openedCount++
        metadata.save()
        auditor.opened(library, book, page())
    }

    @SuppressLint("WrongConstant")
    fun bind(fa: FragmentActivity, pager: ViewPager2) {
        this.pager = pager

        pager.adapter = BookPagerAdapter(fa, book)

        pager.registerOnPageChangeCallback(object : OnPageChangeCallback() {
            override fun onPageSelected(position: Int) {
                bookmark(position)
            }
        })

        pager.offscreenPageLimit = CACHE_PAGES_LIMIT
    }

    fun page(): Int {
        return pager!!.currentItem
    }

    fun page(page: Int) {
        pager!!.setCurrentItem(page, false)
    }

    fun go(change: Int) {
        pager!!.setCurrentItem(pager!!.currentItem + change, false)
    }

    private fun bookmark(page: Int) {
        metadata.lastReadPosition = page
        metadata.save()
    }

    fun resume() {
        page(metadata.lastReadPosition)
    }

    fun finish() {
        if(page() == book.pages - 1) bookmark(0)
        auditor.closed(library, book, page())
    }

    companion object {
        const val CACHE_PAGES_LIMIT = 2
    }
}